mltensor-explainer.md

```diff
@@ -0,0 +1,426 @@
+  const gpuVideoTexture = gpuDevice.importExternalTexture({source: video});
+
+  // Wait for all ML work involving `mlTensor1` to complete, then rent it out to WebGPU.
+  const tensorizedGpuBuffer = await gpuDevice.importExternalBuffer(mlTensor1);
+
+  // Create a bind group for `gpuVideoTexture`, create a command encoder, etc.
+  // to "tensorize" `gpuVideoTexture` and store the result in `tensorizedGpuBuffer`
+  // ...
+
+  gpuDevice.queue.submit([tensorizationCommandEncoder.finish()]);
+
+  // Return the buffer to WebNN.
+  tensorizedGpuBuffer.release();
+
+  // Perform some inference described by `graph` on the frame
+  // (e.g. selfie segmentation)
+  mlContext.dispatch(
+    graph,
+    /*inputs=*/{'input': mlTensor1},
+    /*outputs=*/{'output': mlTensor2},
+  );
+
+  // Wait for all ML work involving `mlTensor2` to complete, then rent it out to WebGPU.
+  const tensorizedGpuBufferAfterInference = await gpuDevice.importExternalBuffer(mlTensor2);
+
+  // Create a bind group for `tensorizedGpuBufferAfterInference`,
+  // create a command encoder, etc to feed `tensorizedGpuBufferAfterInference`
+  // into a GPU shader which may blur the frame or replace background sections
+  // and then render the result
+  // ...
+
+  gpuDevice.queue.submit([texturizeAndRenderCommandEncoder.finish()]);
+
+  // Return the buffer to WebNN for the next frame.
+  tensorizedGpuBufferAfterInference.release();
+
+  // Call this method for each frame.
+  video.requestVideoFrameCallback(applyEffectToFrame);
+}
+```
+
+## Design Discussion
+
+### Timelines
+
+WebNN uses a programming model similar to [WebGPU's](https://www.w3.org/TR/webgpu/#programming-model), in that compute tasks are posted to a timeline - which I've referred to as an "ML context timeline" throughout this document - separate from the content timeline (i.e. "script"). See [the WebGPU documentation of timelines](https://www.w3.org/TR/webgpu/#programming-model-timelines) for more details.
+
+Specifying WebNN timelines is tracked in [#529](https://github.com/webmachinelearning/webnn/issues/529).
+
+### Device Affinity and Relationship to a `GPUDevice`
+
+The WebNN API requires the developer to declare how an `MLTensor` will be used (via `MLTensorUsageFlags`), which the user agent may use as a hint in deciding where to allocate the memory backing an `MLTensor`. Where the memory is ultimately allocated is up to the user agent.
+
+For example [an `MLContext` may be created with a `GPUDevice`](https://www.w3.org/TR/webnn/#dom-ml-createcontext-gpudevice), and creating an `MLTensor` from this context with the `MLTensorUsage.WEBGPU_INTEROP` flag expresses a clear intention to share the tensor with the given `GPUDevice`. However, there is no guarantee that sharing this tensor with WebGPU will be zero-copy.
+
+The `MLTensorUsage.READ` and `MLTensorUsage.WRITE` flags likewise are hints to the user agent indicating that the underlying data will be read and written to, respectively, by script.
```

Comment:
I'm noticing a problem while coding examples with simple usage names like `READ` or `WRITE` - it's ambiguous whether these modifiers are intended for dispatch() or `[read/write]Tensor()` as both can technically read/write the tensor. I think this will lead to confusion.

Since `READ` and `WRITE` usage flags ONLY affect `[read/write]Tensor()`, `READ_TENSOR` and `WRITE_TENSOR` seems like a more appropriate name. @a-sully @fdwr WDYT?

Reply:
**Bryan**: Fine with me. (That was actually one of my original name proposals anyway https://github.com/webmachinelearning/webnn/pull/754#discussion_r1747710638)

Reply:
FYI these properties are now members of `MLTensorDescriptor` as per discussion on https://github.com/webmachinelearning/webnn/pull/754#discussion_r1772300523

| Before    | After |
| -------- | ------- |
| `MLTensorUsage.READ` | `MLTensorDescriptor.readable` |
| `MLTensorUsage.WRITE` | `MLTensorDescriptor.writable` |
| `MLTensorUsage.WEBGPU_INTEROP` | `MLTensorDescriptor.importableToWebGPU` |

This seems more readable than before? Still open to naming suggestions, particularly with regards to the latter member...